### ISSUE TITLE
[Suggestion] Keyboard auto-detect 

### ISSUE THREAD

**Issue by Rstation4** [libretro/RetroArch#12282]
Raspberry Pi 4 B 4gb
Lakka OS Nightly RPI 4 arm Apr 15 2021

Hardware: USB Keyboard and USB Mouse
Cores: DosBox and DosBox-SVN

Im not sure if it is a bug, but when I started a first time setup and attempt at DOSBox the keyboard was not allowing me to input certain keys. IE P was pausing, I quickly had to delete all the keyboard hotkeys, then when Enter and some other keys such as the arrows and x was not inputting I went to controls to change them from gamepad to keyboard and mouse for the core.

It still did not allow me to use those keys until I disconnected the gamepad so RA will not get confused, then using the Keyboard I had to change it from gamepad to keyboard+ mouse and then delete the keys form the input list.

I went ahead and then tested it out on RetroArch (1.9.1) that was installed on my Linux Mint 20.1 Desktop PC. I never had DosBox on it so I downloaded the core and saw that it was doing the same thing there.

**Comment by hizzlekizzle** [comment]
Starting in v1.9.1, there's settings > input > Auto Enable 'Game Focus' Mode (may be an advanced setting, in which case you need to go to settings > user interface > show advanced settings ON before it will be visible), which will automatically trigger the 'game focus mode' setting (mapped to "scroll lock" key by default). This is essentially keyboard/mouse capture, insofar as it sends those events directly to the core instead of listening for hotkeys, etc.

**Comment by MrHuu** [comment]
Yeah, this really left me confused.
With Game Focus disabled, shouldn't all keyboard callbacks by the core be blocked?

The current default behaviour is a half functioning keyboard.
As some buttons still be registerd by the core. And others are blocked even if there's no hotkey assigned to them.

Also, is there an option to disable the on-screen notification when 'Game Focus' is toggled?

**Comment by hizzlekizzle** [comment]
I don't think you can turn just that one off, no. You can turn off widgets and/or "yellow text" notifications in settings > onscreen display > onscreen notifications, though, and save a core override so they only turn off for that core, if you like.

With game focus *not* enabled, the core still listens (i.e., the core is always listening), it just also hits the frontend. Game focus mode shuts the frontend off from listening.

**Comment by MrHuu** [comment]
> 
> 
> I don't think you can turn just that one off, no. You can turn off widgets and/or "yellow text" notifications in settings > onscreen display > onscreen notifications, though, and save a core override so they only turn off for that core, if you like.

To be honest, i've just got GFX widgets working properly on 3DS, and i really like them for achievements and other messages. The widgets take about 1/6 of the screen, so there quite intrusive. But i do not want to disable them total.

Is the 'Game Focus' state exposed outside retroarch.c? Then i'll just add an internal toggle when using the bottom screen as keyboard.

> 
> With game focus _not_ enabled, the core still listens (i.e., the core is always listening), it just also hits the frontend. Game focus mode shuts the frontend off from listening.

For users not familiar with this behavior, default settings will be quite confusing. Since keys which don't have a hotkey assigned to them do get registered by the core. While keys with hotkeys assigned are not.

Let's take the numpad minus as example. Hotkey assigned is volume down. Minus does not get registered by the core. Unless the volume is down to 0%, then it does get registered by dosbox. This is not consisted behavior.

While 'D', which has no hotkey assigned, gets registered by default.
'S' does not, but i can't find out with hotkey or function would be assigned to 'S'. Pressing 'S' just doesn't do nothing by default.

These are just examples, but numerous keys behave in similar fashion. ( talking about windows build )

This, in my opinion, makes the default behavior confusing.
Even for someone who's been using retroarch for years now.

Again, in my opinion, with 'Game Focus' disabled the core should stop listening to all keyboard events.
To create consistent behavior for the entire keyboard.
